### PR TITLE
Vendoring in libnetwork to 3be488927db8d719568917203deddd630a194564

### PR DESCRIPTION
**Pull request by mavenugo**
This PR brings the vendored `libnetwork` code to 3be488927db8d719568917203deddd630a194564, which pulls in quite a few fixes to support libkv, windows daemon compilation fixes, multi-network support for Bridge driver, etc...

Signed-off-by: Madhu Venugopal madhu@docker.com

### CODE REVIEW

**Comment by icecrime** [comment]
Wait, this vendors Swarm into Docker, and Swarm already vendors Docker.

image

**Comment by mavenugo** [comment]
:) @icecrime `libnetwork` uses swarm packages `discovery` and `store` as libraries and unlike Godeps, I couldn't selectively pick those packages alone in the vendor.sh ...  and hence it shows up like a circular dependancy. Is there a way to just pull in the required packages instead of the entire project ?

**Comment by icecrime** [comment]
Intuitively, and without looking into more details, I'm tempted to say that `discovery` and `store` should be pulled out of Swarm as a common dependency to both Swarm and libnetwork, but I'm sure if it was that easy it would already be the case :-/

**Comment by mavenugo** [comment]
@icecrime yes, we started to work as an independent project for the store, but decided to pull it into swarm for execution reasons. We can pull it out. am not too sure about `discovery`. @vieux @aluzzardi do you have any comments ?

**Comment by LK4D4** [comment]
It would be cool if we have more nice libraries under docker/ account.

**Comment by mavenugo** [comment]
@icecrime thanks to the distribution vendoring process... followed a similar approach of removing unused packages. Now the dependencies are narrowed to just those packages that libnetwork is dependent on (discovery, store, consul/etcd/zookeeper client apis).

File: hack/vendor.sh [hunk 1]
```diff
@@ -55,9 +55,39 @@ clone hg code.google.com/p/go.net 84a4013f96e0
...
+mkdir -p src/github.com/docker/swarm/pkg
+mv tmp-store src/github.com/docker/swarm/pkg/store
+
+clone git github.com/hashicorp/consul 954aec66231b79c161a4122b023fbcad13047f79
+mv src/github.com/hashicorp/consul/api tmp-api
+rm -rf src/github.com/hashicorp/consul
+mkdir -p src/github.com/hashicorp/consul
+mv tmp-api src/github.com/hashicorp/consul/api
+
+clone git github.com/coreos/go-etcd 73a8ef737e8ea002281a28b4cb92a1de121ad4c6
```
Comment (tianon) [inline hack/vendor.sh]: Can we use a tagged release of `go-etcd` instead? (https://github.com/coreos/go-etcd/releases)

Same with consul (https://github.com/hashicorp/consul/releases).

Comment (tianon) [inline hack/vendor.sh]: (and if they don't have a recent enough release, we've found that most upstreams are pretty reasonable about tagging a new one)

Comment (LK4D4) [inline hack/vendor.sh]: looks like we need func for such vendoring :)

File: hack/vendor.sh [hunk 2]
```diff
@@ -55,9 +55,39 @@ clone hg code.google.com/p/go.net 84a4013f96e0
...
 #get libnetwork packages
-clone git github.com/docker/libnetwork 005bc475ee49a36ef2ad9c112d1b5ccdaba277d4
+clone git github.com/docker/libnetwork a09a60a123b4106954e5b42ed5ec41f264eeb8bf
 clone git github.com/vishvananda/netns 008d17ae001344769b031375bdb38a86219154c6
 clone git github.com/vishvananda/netlink 8eb64238879fed52fd51c5b30ad20b928fb4c36c
+clone git github.com/BurntSushi/toml f706d00e3de6abe700c994cdd545a1a4915af060
+clone git github.com/deckarep/golang-set ef32fa3046d9f249d399f98ebaf9be944430fd1d
+
+#libnetwork uses swarm discovery and store libraries
+clone git github.com/docker/swarm 54dfabd2521314de1c5b036f6c609efbe09df4ea
```
Comment (LK4D4) [inline hack/vendor.sh]: looks like we need func for such stuff :)

File: hack/vendor.sh (hunk 1 above)
Comment (mavenugo) [inline hack/vendor.sh]: @tianon good point. I will cross-check with swarm discovery/store to make sure the selected tag satisfies the dependency needs.

File: hack/vendor.sh (hunk 2 above)
Comment (mavenugo) [inline hack/vendor.sh]: @LK4D4 @tianon if we are okay with this approach, then yes, having a generic function to filter out unused packages & pulling in only the necessary packages will be helpful.

**Comment by abronan** [comment]
@mavenugo @icecrime We will pull `pkg/store` out in its own repository but not sure this is going to happen before 1.7.0. If you found a way to vendor only the relevant parts then it's all good.

We can try to pull out `discovery` the same way if this is more convenient for `libnetwork` (we'd like to do this in Swarm as well I guess but this is a bit more difficult as it is more tied with Swarm).

**Comment by icecrime** [comment]
* I don't think we have any dependency on 1.7.0 there
* I'm +1 for pulling out rather than import from swarm, but that's just my 2cts

**Comment by calavera** [comment]
Do we have any resolution to move this forward?

@icecrime the vendor script actually removes all the pieces that we don't need from swarm. I agree that's not ideal, but it might not be too bad in the short term.

@abronan is there anything I can help to extract that discovery package? Is it just a matter of extracting that directory to a new repo and updating dependencies, or we should take something else into account?

**Comment by icecrime** [comment]
@calavera I think it has been taken care of (circular dependency was broken), but that the  PR was not yet updated.

**Comment by mavenugo** [comment]
Yes. I will push an update today.

**Comment by mavenugo** [comment]
@icecrime @calavera @abronan migrated to libkv and addressed other comments. PTAL.

**Comment by mavenugo** [comment]
@jhowardmsft FYI. pls check & confirm if the windows compilation issues related to libnetwork are addressed.

**Comment by aluzzardi** [comment]
This doesn't load on my browser :) Just to check, are you now referencing to `docker/libkv` instead of `docker/swarm/pkg/store`? We put it in a separate repo a few days ago

**Comment by mavenugo** [comment]
@aluzzardi yes. thanks for making that happen. appreciate it.

**Comment by calavera** [comment]
It looks like we're vendoring a testing library that we don't need. I think https://github.com/docker/libkv/pull/6 should fix this problem.

**Comment by mavenugo** [comment]
@calavera yes.

**Comment by jhowardmsft** [comment]
@mavenugo  lgtm
LGTM (from a Windows compile point of view) :+1:

**Comment by mavenugo** [comment]
Thanks @jhowardmsft

File: hack/vendor.sh [hunk 3]
```diff
@@ -17,9 +17,16 @@ clone hg code.google.com/p/go.net 84a4013f96e0
 clone hg code.google.com/p/gosqlite 74691fb6f837
 
 #get libnetwork packages
-clone git github.com/docker/libnetwork e578e95aa101441481411ff1d620f343895f24fe
+clone git github.com/docker/libnetwork 3be488927db8d719568917203deddd630a194564
+clone git github.com/docker/libkv ab16c3d4a8785a9877c62d0b11ea4441cf09120c
+clone git github.com/stretchr/testify dab07ac62d4905d3e48d17dc549c684ac3b7c15a
```
Comment (tianon) [inline hack/vendor.sh]: Can we use `v1.0` here instead? (https://github.com/stretchr/testify/releases)

**Comment by mavenugo** [comment]
Thanks @calavera I will pick up a8d22fa21a3b4626aa33364ca2603c84f2622c32 from libkv.

**Comment by calavera** [comment]
@mavenugo sorry, you need docker/libkv#7. There was some confusion about the right fix.

**Comment by mavenugo** [comment]
@calavera @abronan its a good fix. but unfortunately libkv uses the mock package from testify that pulls in the deps back in. seems like mock is a service provided by libkv. :(

**Comment by mavenugo** [comment]
Thanks @abronan @aluzzardi for offering to remove the mock from libkv top level provider. That will surely get rid of testify.

**Comment by mavenugo** [comment]
@aluzzardi @abronan Thanks for the quick work. testify is now removed.
@calavera @icecrime PTAL

**Comment by icecrime** [comment]
Ping @tianon.

**Comment by tianon** [comment]
LGTM :+1:

**Comment by tianon** [comment]
I'm very curious what libnetwork uses TOML internally for though. :smile:

**Comment by mavenugo** [comment]
@tianon Thanks for the review. libnetwork provides a tool named dnet, which uses TOML for configuration.

**Comment by icecrime** [comment]
LGTM! Thanks for extracting the libs <3

**Pull request merged**